mod/state-transition/pkg/core/state_processor_staking.go

```diff
@@ -308,24 +299,26 @@ func (sp *StateProcessor[
 		return err
 	}
 
-	// Update the next validator index to start the next withdrawal sweep
+	// Update the next validator index to start the next withdrawal sweep.
 	//#nosec:G701 // won't overflow in practice.
 	if numWithdrawals == int(sp.cs.MaxWithdrawalsPerPayload()) {
-		// Next sweep starts after the latest withdrawal's validator index
+		// Next sweep starts after the latest withdrawal's validator index.
 		nextValidatorIndex =
-			(expectedWithdrawals[len(expectedWithdrawals)-1].GetIndex() + 1) %
-				math.U64(totalValidators)
-	} else {
-		// Advance sweep by the max length of the sweep if there was not
-		// a full set of withdrawals
+			(expectedWithdrawals[numWithdrawals-1].GetValidatorIndex() + 1) %
+				math.ValidatorIndex(totalValidators)
+		return st.SetNextWithdrawalValidatorIndex(nextValidatorIndex)
+	} else if numWithdrawals > 1 {
+		// Advance sweep by the max length of the sweep if there was not a full
+		// set of withdrawals.
 		nextValidatorIndex, err = st.GetNextWithdrawalValidatorIndex()
 		if err != nil {
 			return err
 		}
 		nextValidatorIndex += math.ValidatorIndex(
 			sp.cs.MaxValidatorsPerWithdrawalsSweep())
 		nextValidatorIndex %= math.ValidatorIndex(totalValidators)
+		return st.SetNextWithdrawalValidatorIndex(nextValidatorIndex)
 	}
 
-	return st.SetNextWithdrawalValidatorIndex(nextValidatorIndex)
+	return nil
```

Comment:
If `numWithdrawals == 0`, should that not be an error as we expected to have always the EVM inflation withdrawal first?

Reply:
Agree that we should error if numWithdrawals == 0.